standalone/pruntime/enclave/src/system/gk.rs

```diff
@@ -437,9 +459,662 @@ mod tokenomic {
+pub mod tests {
+    use super::Gatekeeper;
+
+    use super::msg_trait::{EgressMessage, MessageChannel};
+    use super::tokenomic::fp;
+    use super::BlockInfo;
+    use crate::std::cell::RefCell;
+    use crate::std::vec::Vec;
+    use parity_scale_codec::Encode;
+    use phala_mq::{BindTopic, Message, MessageDispatcher, MessageOrigin};
+    use phala_types::{messaging as msg, WorkerPublicKey};
+
+    trait DispatcherExt {
+        fn dispatch_bound<M: Encode + BindTopic>(&mut self, sender: &MessageOrigin, msg: M);
+    }
+
+    impl DispatcherExt for MessageDispatcher {
+        fn dispatch_bound<M: Encode + BindTopic>(&mut self, sender: &MessageOrigin, msg: M) {
+            let _ = self.dispatch(mk_msg(sender, msg));
+        }
+    }
+
+    fn mk_msg<M: Encode + BindTopic>(sender: &MessageOrigin, msg: M) -> Message {
+        Message {
+            sender: sender.clone(),
+            destination: M::TOPIC.to_vec().into(),
+            payload: msg.encode(),
+        }
+    }
+
+    #[derive(Default)]
+    struct CollectChannel {
+        messages: RefCell<Vec<EgressMessage>>,
+    }
+
+    impl MessageChannel for CollectChannel {
+        fn push_message(&self, message: EgressMessage) {
+            self.messages.borrow_mut().push(message);
+        }
+    }
+
+    struct Roles {
+        mq: MessageDispatcher,
+        gk: Gatekeeper<CollectChannel>,
+        workers: [WorkerPublicKey; 2],
+    }
+
+    impl Roles {
+        fn test_roles() -> Roles {
+            let mut mq = MessageDispatcher::new();
+            let egress = CollectChannel::default();
+            let gk = Gatekeeper::new(&mut mq, egress);
+            Roles {
+                mq,
+                gk,
+                workers: [
+                    WorkerPublicKey::from_raw([0x01u8; 33]),
+                    WorkerPublicKey::from_raw([0x02u8; 33]),
+                ],
+            }
+        }
+
+        fn for_worker(&mut self, n: usize) -> ForWorker {
+            ForWorker {
+                mq: &mut self.mq,
+                pubkey: &self.workers[n],
+            }
+        }
+    }
+
+    struct ForWorker<'a> {
+        mq: &'a mut MessageDispatcher,
+        pubkey: &'a WorkerPublicKey,
+    }
+
+    impl ForWorker<'_> {
+        fn pallet_say(&mut self, event: msg::WorkerEvent) {
+            let sender = MessageOrigin::Pallet(b"Pallet".to_vec());
+            let message = msg::SystemEvent::new_worker_event(self.pubkey.clone(), event);
+            self.mq.dispatch_bound(&sender, message);
+        }
+
+        fn say<M: Encode + BindTopic>(&mut self, message: M) {
+            let sender = MessageOrigin::Worker(self.pubkey.clone());
+            self.mq.dispatch_bound(&sender, message);
+        }
+
+        fn challenge(&mut self) {
+            use sp_core::U256;
+
+            let sender = MessageOrigin::Pallet(b"Pallet".to_vec());
+            let pkh = sp_core::blake2_256(self.pubkey.as_ref());
+            let hashed_id: U256 = pkh.into();
```

Comment:
Not necessary to hash. Any 256bits random number should work

Reply:
Hash it in order to filter out the other workers.

Reply:
Ah, brilliant! Could you add a comment for this?

Reply:
Sure, done.